acme/acme/mixins.py

```diff
@@ -0,0 +1,55 @@
+"""Useful mixins for Challenge and Resource objects"""
+
+
+class VersionedLEACMEMixin(object):
+    """This mixin allows to store the current ACME version as a property"""
+    @property
+    def le_auto_version(self):
+        """Define the version of ACME protocol to use"""
+        return getattr(self, '_le_auto_version', 1)
+
+    @le_auto_version.setter
+    def le_auto_version(self, version):
+        # We need to use object.__setattr__ to not depend on the specific implementation of
+        # __setattr__  in current class (eg. jose.TypedJSONObjectWithFields raises AttributeError
+        # for any attempt to set an attribute to make objects immutable).
+        object.__setattr__(self, '_le_auto_version', version)
+
+    def __setattr__(self, key, value):
+        if key == 'le_auto_version':
+            # Required for @property to operate properly. See comment above.
+            object.__setattr__(self, key, value)
+        else:
+            super(VersionedLEACMEMixin, self).__setattr__(key, value)  # pragma: no cover
+
+
+class ResourceMixin(VersionedLEACMEMixin):
+    """
+    This mixin allows to generate a RFC8555 compliant JWS payload
+    by removing the `resource` field if needed (eg. ACME v2 protocol).
+    """
+    def to_partial_json(self):
```

Comment:
nit: While maybe not intentionally part of the public API `josepy.JSONDeserializable` defines the method `fields_to_partial_json`. In my comment at https://github.com/certbot/certbot/pull/7176#issuecomment-567718597, I suggested we overrode `fields_to_partial_json` since `to_partial_json` calls it, but that's making assumptions about the internals of  `josepy.JSONDeserializable.to_partial_json`.

Instead, what do you think about overriding both methods here and ensuring `resource` is popped in both cases. Bonus points for any code reuse if it's sane to do so.

Reply:
I tried to do so, but modifying `fields_to_partial_json` has a lot more implications. In particular, every call that implies `josepy.JSONDeserializable._check_required()` will fail because `resource` or `type` is missing in the generated json object while it is defined as a property at class level.

Since it is sufficient for the purpose of this PR (fixing compliance in regards to the ACME server), I would like to avoid to modify `fields_to_partial_json` and avoid to add more workarounds here.

Reply:
> In particular, every call that implies `josepy.JSONDeserializable._check_required()` will fail because `resource` or `type` is missing in the generated json object while it is defined as a property at class level.

Is this the problem? I believe `_check_required` is only called when converting JSON to the class and not when dumping the class to JSON.

Playing with your code before the revert commit, I was able to get all unit tests to pass by fixing a small typo:
```
diff --git a/acme/acme/mixins.py b/acme/acme/mixins.py
index d4d78636c..1d76280c1 100644
--- a/acme/acme/mixins.py
+++ b/acme/acme/mixins.py
@@ -58,7 +58,7 @@ class TypeMixin(VersionedLEACMEMixin):
 def _safe_jobj_compliance(instance, jobj_method, uncompliant_field):
     if hasattr(instance, jobj_method):
         jobj = getattr(instance, jobj_method)()
-        if instance.le_acme_version:
+        if instance.le_acme_version == 2:
             jobj.pop(uncompliant_field, None)
         return jobj
```
If you still don't like this approach for some reason though, another thing we could do which would satisfy me is to deprecate `fields_to_partial_json` in `josepy` by having the method send a deprecation warning. Probably the easiest way to do that is to just make a private `_fields_to_partial_json` method which everything else (including the deprecated `fields_to_partial_json` method) can still call.

My overall concern with my comment here is just leaving the problems fixed by this PR unresolved for some methods on the objects. Removing the problematic methods also fixes that. If we take this approach though, I'd kind of like the `josepy` deprecation to land before we land this PR to make sure we actually do that work.

Reply:
You are right. I misinterpretated the error.